### PR TITLE
more adjustments to work with new sane add plugin

### PR DESCRIPTION
**Pull request by czpython**
* Fixed backwards compatiblity with projects that already have "ghost" text plugins.
* Canceling text plugin now correctly removes the cmsplugin record.
* Canceling a child plugin of the text plugin correctly removes the child plugin.

### CODE REVIEW

**Comment by czpython** [comment]
@vxsx Things to think about:

* Security. When deleting a plugin through the cancel endpoint, we check the add permissions because a user shouldn't need delete permissions to add -> cancel plugin.
  This made this even more difficult because it means the endpoint needs to somehow prevent users from deleting any plugin. Even if we were to check if the plugin is text only or a child of a text plugin, it still poses an issue to just allow users to delete like this.
* It gets very difficult to return the cancel url and token on the widget template.
  Also I don't think is good idea because the widget itself was made to work outside of the pages setting.
* When adding child plugins to ckeditor, we have no control over the add logic because the CMS calls the add view of the respective plugin class. This means we can't return the custom data attributes like plugin cancel endpoint and cancel token.


Solutions (up to discussion):

* I've added a cancel token which is unique per plugin type and plugin id.
  This token prevents users from hitting the cancel endpoint with any plugin id.
  Instead, the given plugin id and token need to match.
  The token itself is not stored anywhere, it just leverages django's [signing](https://docs.djangoproject.com/ja/1.9/topics/signing/#using-the-low-level-api).
*  In order to provide the cancel url and cancel token when adding a text plugin, I suggest we switch the add response of the add-plugin endpoint to be a JSON response with ``edit-url``, ``cancel-url`` and ``cancel-token``. This would then guarantee that the cancel token and url only show up after adding the plugin vs using a GET parameter.
*  This is the most challenging one, I think the only way to truly solve this without adding logic to the CMS is to have a dedicated endpoint for adding plugins on ckeditor that wraps the cms endpoint (logic wise) and takes care of generating and passing the cancel token and url.

Also @ojii and I have been making the changes here to be backwards compatible with other CMS versions (not only sane-add-plugin version), are the frontend changes like this as well? :8ball:

**Comment by vxsx** [comment]
@czpython these are not backwards compatible

**Comment by czpython** [comment]
@vxsx I've added a new flag to the ckeditor config called ``delete_on_cancel``.
When this flag is true, send the POST request on cancel for the text plugin only.
This flag does not affect child plugins.

The cancel_token is to be used whenever a user cancels adding a child plugin.

**Comment by vxsx** [comment]
after my adjustments it kinda works, but i hit this constantly when trying to remove multiple plugins
can reproduce consistently when cancelling fresh text plugin with label
```
OperationalError at /en/admin/cms/page/plugin/text_plugin/delete-on-cancel/
database is locked

Request Method: POST
Request URL: http://localhost:8000/en/admin/cms/page/plugin/text_plugin/delete-on-cancel/
Django Version: 1.8.12
Python Executable: /Users/divio/work/django-cms/env/bin/python
Python Version: 2.7.8
Python Path: [u'.', '/Users/divio/work/django-cms', '/Users/divio/work/django-cms/env/src/djangocms-column', '/Users/divio/work/django-cms/env/src/djangocms-style', '/Users/divio/work/django-cms/env/src/djangocms-file', '/Users/divio/work/django-cms/env/src/djangocms-googlemap', '/Users/divio/work/django-cms/env/src/djangocms-inherit', '/Users/divio/work/django-cms/env/src/djangocms-picture', '/Users/divio/work/django-cms/env/src/djangocms-teaser', '/Users/divio/work/django-cms/env/src/djangocms-video', '/Users/divio/work/django-cms/env/src/djangocms-link', '/Users/divio/work/djangocms-text-ckeditor', '/Users/divio/work/django-cms/env/lib/python27.zip', '/Users/divio/work/django-cms/env/lib/python2.7', '/Users/divio/work/django-cms/env/lib/python2.7/plat-darwin', '/Users/divio/work/django-cms/env/lib/python2.7/plat-mac', '/Users/divio/work/django-cms/env/lib/python2.7/plat-mac/lib-scriptpackages', '/Users/divio/work/django-cms/env/lib/python2.7/lib-tk', '/Users/divio/work/django-cms/env/lib/python2.7/lib-old', '/Users/divio/work/django-cms/env/lib/python2.7/lib-dynload', '/usr/local/Cellar/python/2.7.8_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7', '/usr/local/Cellar/python/2.7.8_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-darwin', '/usr/local/Cellar/python/2.7.8_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-tk', '/usr/local/Cellar/python/2.7.8_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-mac', '/usr/local/Cellar/python/2.7.8_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/plat-mac/lib-scriptpackages', '/Users/divio/work/django-cms/env/lib/python2.7/site-packages']
Server time: Wed, 13 Apr 2016 04:27:42 -0500
Installed Applications:
[u'django.contrib.contenttypes',
 u'django.contrib.auth',
 u'django.contrib.sessions',
 u'django.contrib.sites',
 u'django.contrib.staticfiles',
 u'djangocms_admin_style',
 u'django.contrib.admin',
 u'djangocms_helper.test_data',
 u'django.contrib.messages',
 u'cms',
 u'menus',
 u'sekizai',
 u'reversion',
 u'djangocms_text_ckeditor',
 u'djangocms_grid',
 u'filer',
 u'aldryn_bootstrap3',
 u'cms.test_utils.project.placeholderapp',
 u'treebeard',
 u'mptt']
Installed Middleware:
[u'django.middleware.http.ConditionalGetMiddleware',
 u'django.contrib.sessions.middleware.SessionMiddleware',
 u'django.contrib.auth.middleware.AuthenticationMiddleware',
 u'django.contrib.messages.middleware.MessageMiddleware',
 u'django.middleware.csrf.CsrfViewMiddleware',
 u'django.middleware.locale.LocaleMiddleware',
 u'django.middleware.common.CommonMiddleware',
 u'cms.middleware.language.LanguageCookieMiddleware',
 u'cms.middleware.user.CurrentUserMiddleware',
 u'cms.middleware.page.CurrentPageMiddleware',
 u'cms.middleware.toolbar.ToolbarMiddleware',
 u'cms.middleware.utils.ApphookReloadMiddleware']

Traceback:
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/Users/divio/work/djangocms-text-ckeditor/djangocms_text_ckeditor/cms_plugins.py" in delete_on_cancel
  190.             plugin.delete()
File "/Users/divio/work/django-cms/cms/models/pluginmodel.py" in delete
  479.             super(CMSPlugin, self).delete(*args, **kwargs)
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/treebeard/models.py" in delete
  506.         self.__class__.objects.filter(pk=self.pk).delete()
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/treebeard/mp_tree.py" in delete
  98.             super(MP_NodeQuerySet, qset).delete()
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/models/query.py" in delete
  537.         collector.delete()
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/models/deletion.py" in delete
  287.                         sender=model, instance=obj, using=self.using
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/dispatch/dispatcher.py" in send
  189.             response = receiver(signal=self, sender=sender, **named)
File "/Users/divio/work/django-cms/cms/signals/plugins.py" in pre_delete_plugins
  57.     set_dirty(plugin)
File "/Users/divio/work/django-cms/cms/signals/plugins.py" in set_dirty
  29.             Title.objects.filter(page=placeholder.page, language=language).update(publisher_state=PUBLISHER_STATE_DIRTY)
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/models/query.py" in update
  563.             rows = query.get_compiler(self.db).execute_sql(CURSOR)
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/models/sql/compiler.py" in execute_sql
  1062.         cursor = super(SQLUpdateCompiler, self).execute_sql(result_type)
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/models/sql/compiler.py" in execute_sql
  840.             cursor.execute(sql, params)
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  79.             return super(CursorDebugWrapper, self).execute(sql, params)
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  64.                 return self.cursor.execute(sql, params)
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/utils.py" in __exit__
  98.                 six.reraise(dj_exc_type, dj_exc_value, traceback)
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  64.                 return self.cursor.execute(sql, params)
File "/Users/divio/work/django-cms/env/lib/python2.7/site-packages/django/db/backends/sqlite3/base.py" in execute
  318.         return Database.Cursor.execute(self, query, params)

Exception Type: OperationalError at /en/admin/cms/page/plugin/text_plugin/delete-on-cancel/
Exception Value: database is locked
Request information:
GET: No GET data

POST:
token = u'TziUXZYN9N0p2x2gHQDiLdTFxoo'
plugin = u'11'

FILES: No FILES data

COOKIES:
djangocms_nodes_open = 'page_1%2Cpage_4%2Cpage_5'
askbot_visitor = 'False'
addevent_track_cookie = '17aec3e7-5dc5-45e8-bb86-5196460b312c'
lc_window_state = 'minimized'
__lc.visitor_id.4680011 = 'S1433334318.05d5d7c901'
__atuvc = '3%7C4'
snatched_view = 'list'
pageview_visit = '133'
wanted_view = 'thumb'
suggest_view = 'thumb'
current_section = 'residential'
intercom-visitor-semaphore-wcfe7111 = '1'
imdb_box_office_view = 'thumb'
s_cc = 'true'
ajs_anonymous_id = '%225848b29b-4c4c-4340-aa4f-13bc0f414af7%22'
late_view = 'list'
advanced_toggle_checked = '0'
gsScrollPos = '0'
ajs_user_id = 'null'
s_lv = '1429996952625'
_ga = 'GA1.1.1738864788.1429972332'
sessionid = 'pv4jg5gg5w7hob6704ng2p84zp7lq3zz'
s_sq = '%5B%5BB%5D%5D'
csrftoken = 'oGs1aPcCoJgn4AETkCY7RTmcJ0351AWL'
_hp2_id.312726686 = '7278556461588913.3673023101.2391915934'
mp_dfda6122b4e5c8ee8d4d773bc4a3c71f_mixpanel = '%7B%22distinct_id%22%3A%20%2214ed4abd64f7c4-0063d0e7b-33677e01-232800-14ed4abd650e37%22%2C%22%24initial_referrer%22%3A%20%22http%3A%2F%2Flocalhost%3A5000%2F%22%2C%22%24initial_referring_domain%22%3A%20%22localhost%3A5000%22%7D'
soon_view = 'thumb'
blu_ray_com_new_releases_view = 'thumb'
ajs_group_id = 'null'
imdb_top_dvd_rentals_view = 'thumb'
django_language = 'en'
s_fid = '7FC16419AEE876E2-0634AD3C567D8DD5'

META:
RUN_MAIN = 'true'
HTTP_REFERER = 'http://localhost:8000/en/admin/cms/page/edit-plugin/11/?delete-on-cancel'
LESS = '-R'
SERVER_SOFTWARE = 'WSGIServer/0.1 Python/2.7.8'
LC_CTYPE = 'UTF-8'
SCRIPT_NAME = u''
REQUEST_METHOD = 'POST'
ZSH_CUSTOM = '/Users/divio/.dotfiles/zsh/'
HTTP_ORIGIN = 'http://localhost:8000'
SERVER_PROTOCOL = 'HTTP/1.1'
VIRTUALENVWRAPPER_SCRIPT = '/usr/local/bin/virtualenvwrapper.sh'
CONTENT_LENGTH = '43'
VIRTUAL_ENV = '/Users/divio/work/django-cms/env'
SHELL = '/bin/zsh'
CSRF_COOKIE = u'oGs1aPcCoJgn4AETkCY7RTmcJ0351AWL'
SECURITYSESSIONID = '186a7'
XPC_FLAGS = '0x0'
NVM_BIN = '/Users/divio/.nvm/versions/node/v5.0.0/bin'
PIP_REQUIRE_VIRTUALENV = 'true'
VIRTUAL_ENV_DISABLE_PROMPT = '1'
MANPATH = '/Users/divio/.nvm/versions/node/v5.0.0/share/man:/usr/share/man:/usr/local/share/man:/opt/X11/share/man:/Applications/Xcode.app/Contents/Developer/usr/share/man:/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/share/man'
SVN_EDITOR = 'vim'
ARCHFLAGS = '-arch x86_64'
WORKON_HOME = '/Users/divio/.virtualenvs'
HTTP_PRAGMA = 'no-cache'
HTTP_X_REQUESTED_WITH = 'XMLHttpRequest'
HTTP_CACHE_CONTROL = 'no-cache'
HTTP_ACCEPT = '*/*'
wsgi.version = 
wsgi.multiprocess = False
DATABASE_URL = 'sqlite://localhost/testdb.sqlite'
HTTP_COOKIE = 'djangocms_nodes_open=page_1%2Cpage_4%2Cpage_5; gsScrollPos=0; addevent_track_cookie=17aec3e7-5dc5-45e8-bb86-5196460b312c; askbot_visitor=False; pageview_visit=133; s_cc=true; s_fid=7FC16419AEE876E2-0634AD3C567D8DD5; s_lv=1429996952625; s_sq=%5B%5BB%5D%5D; __lc.visitor_id.4680011=S1433334318.05d5d7c901; lc_window_state=minimized; ajs_anonymous_id=%225848b29b-4c4c-4340-aa4f-13bc0f414af7%22; mp_dfda6122b4e5c8ee8d4d773bc4a3c71f_mixpanel=%7B%22distinct_id%22%3A%20%2214ed4abd64f7c4-0063d0e7b-33677e01-232800-14ed4abd650e37%22%2C%22%24initial_referrer%22%3A%20%22http%3A%2F%2Flocalhost%3A5000%2F%22%2C%22%24initial_referring_domain%22%3A%20%22localhost%3A5000%22%7D; _hp2_id.312726686=7278556461588913.3673023101.2391915934; advanced_toggle_checked=0; current_section=residential; __atuvc=3%7C4; wanted_view=thumb; snatched_view=list; soon_view=thumb; suggest_view=thumb; late_view=list; blu_ray_com_new_releases_view=thumb; imdb_box_office_view=thumb; imdb_top_dvd_rentals_view=thumb; ajs_user_id=null; ajs_group_id=null; intercom-visitor-semaphore-wcfe7111=1; _ga=GA1.1.1738864788.1429972332; django_language=en; sessionid=pv4jg5gg5w7hob6704ng2p84zp7lq3zz; csrftoken=oGs1aPcCoJgn4AETkCY7RTmcJ0351AWL'
PATH_INFO = u'/en/admin/cms/page/plugin/text_plugin/delete-on-cancel/'
QUERY_STRING = ''
NVM_IOJS_ORG_MIRROR = 'https://iojs.org/dist'
HTTP_USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/49.0.2623.112 Safari/537.36'
HTTP_CONNECTION = 'keep-alive'
REMOTE_ADDR = '127.0.0.1'
SHLVL = '3'
wsgi.url_scheme = 'http'
EDITOR = '/Applications/MacVim.app/Contents/MacOS/Vim'
GPG_AGENT_INFO = '/tmp/gpg-zxlYMk/S.gpg-agent:59268:1'
XPC_SERVICE_NAME = '0'
TMUX = '/private/var/folders/qf/m9l_j5_j3gz9vzfc3vkdj5300000gn/T/tmux-501/default,34206,1'
wsgi.multithread = True
ITERM_PROFILE = 'Default'
_ = '/Users/divio/work/django-cms/env/bin/python'
wsgi.file_wrapper = ''
REMOTE_HOST = ''
HTTP_ACCEPT_ENCODING = 'gzip, deflate'
HTTP_X_CSRFTOKEN = 'oGs1aPcCoJgn4AETkCY7RTmcJ0351AWL'
NVM_NODEJS_ORG_MIRROR = 'https://nodejs.org/dist'
NVM_DIR = '/Users/divio/.nvm'
HOME = '/Users/divio'
DISPLAY = '/private/tmp/com.apple.launchd.TPNk49S7bO/org.macosforge.xquartz:0'
TERM_PROGRAM = 'iTerm.app'
Apple_PubSub_Socket_Render = '/private/tmp/com.apple.launchd.WclYmg4SLx/Render'
VIRTUALENVWRAPPER_PYTHON = '/usr/local/bin/python'
NVM_PATH = '/Users/divio/.nvm/versions/node/v5.0.0/lib/node'
TMPDIR = '/var/folders/qf/m9l_j5_j3gz9vzfc3vkdj5300000gn/T/'
SERVER_PORT = '8000'
HTTP_HOST = 'localhost:8000'
SOLARIZED_THEME = 'light'
ZSH_THEME = 'vxsx'
VIRTUALENVWRAPPER_HOOK_DIR = '/Users/divio/.virtualenvs'
wsgi.run_once = False
wsgi.errors = <open file '<stderr>', mode 'w' at 0x1014891e0>
HTTP_ACCEPT_LANGUAGE = 'en-US,en;q=0.8,fr;q=0.6,ru;q=0.4,uk;q=0.2,de;q=0.2'
__CF_USER_TEXT_ENCODING = '0x1F5:0x0:0x0'
TMUX_PANE = '%37'
USER = 'divio'
LOGNAME = 'divio'
PATH = '/Users/divio/work/django-cms/env/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/divio/.nvm/versions/node/v5.0.0/bin:/opt/local/bin:/opt/local/sbin:/Users/divio/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/usr/local/sbin:/opt/local/bin:/opt/local/sbin:/Users/divio/bin:/Users/divio/.rvm/bin:/Users/divio/Library/Python/2.7/bin:/Users/divio/.local/bin:/Users/divio/.rvm/bin:/Users/divio/Library/Python/2.7/bin:/Users/divio/.local/bin'
ZSH = '/Users/divio/.oh-my-zsh'
TERM = 'xterm-256color-italic'
SERVER_NAME = '1.0.0.127.in-addr.arpa'
ITERM_SESSION_ID = 'w0t0p1:58F1EC06-B724-4B75-801E-BB293E290BB5'
SSH_AUTH_SOCK = '/tmp/ssh-agent-divio-screen'
wsgi.input = <socket._fileobject object at 0x103c26f50>
LC_ALL = 'en_US.UTF-8'
VIRTUALENVWRAPPER_PROJECT_FILENAME = '.project'
LSCOLORS = 'Gxfxcxdxbxegedabagacad'
GATEWAY_INTERFACE = 'CGI/1.1'
PAGER = 'less'
OLDPWD = '/Users/divio/work/django-cms'
PWD = '/Users/divio/work/django-cms'
PIP_DOWNLOAD_CACHE = '/Users/divio/.pip/cache'
CONTENT_TYPE = 'application/x-www-form-urlencoded; charset=UTF-8'

Settings:
Using settings module 
SECURE_BROWSER_XSS_FILTER = False
USE_THOUSAND_SEPARATOR = False
CSRF_COOKIE_SECURE = False
LANGUAGE_CODE = u'en'
ROOT_URLCONF = u'djangocms_helper.urls'
MANAGERS = 
SILENCED_SYSTEM_CHECKS = []
CMS_TOOLBAR_URL__EDIT_ON = u'edit'
SESSION_SERIALIZER = 'django.contrib.sessions.serializers.JSONSerializer'
STATIC_ROOT = u'/var/folders/qf/m9l_j5_j3gz9vzfc3vkdj5300000gn/T/tmp7TCLMjstatic'
ALLOWED_HOSTS = []
MESSAGE_STORAGE = 'django.contrib.messages.storage.fallback.FallbackStorage'
EMAIL_SUBJECT_PREFIX = '[Django] '
SERVER_EMAIL = 'root@localhost'
SECURE_HSTS_SECONDS = 0
STATICFILES_FINDERS = [u'django.contrib.staticfiles.finders.FileSystemFinder', u'django.contrib.staticfiles.finders.AppDirectoriesFinder']
SESSION_CACHE_ALIAS = 'default'
SESSION_COOKIE_DOMAIN = None
SESSION_COOKIE_NAME = 'sessionid'
TIME_INPUT_FORMATS = 
SECURE_REDIRECT_EXEMPT = []
DATABASES = {u'default': {'ENGINE': 'django.db.backends.sqlite3', 'AUTOCOMMIT': True, 'ATOMIC_REQUESTS': False, 'NAME': 'testdb.sqlite', 'CONN_MAX_AGE': 0, 'TIME_ZONE': 'UTC', 'PORT': '', 'HOST': 'localhost', 'USER': '', 'TEST': {'COLLATION': None, 'CHARSET': None, 'NAME': None, 'MIRROR': None}, 'PASSWORD': u'********************', 'OPTIONS': {}}}
EMAIL_SSL_KEYFILE = u'********************'
FILE_UPLOAD_DIRECTORY_PERMISSIONS = None
FILE_UPLOAD_PERMISSIONS = None
FILE_UPLOAD_HANDLERS = 
DEFAULT_CONTENT_TYPE = 'text/html'
APPEND_SLASH = True
LOCALE_PATHS = 
DATABASE_ROUTERS = []
DEFAULT_TABLESPACE = ''
YEAR_MONTH_FORMAT = 'F Y'
STATICFILES_STORAGE = 'django.contrib.staticfiles.storage.StaticFilesStorage'
CACHES = {'default': {'BACKEND': 'django.core.cache.backends.locmem.LocMemCache'}}
SESSION_COOKIE_PATH = '/'
SECURE_CONTENT_TYPE_NOSNIFF = False
CMS_TEMPLATES = 
MIDDLEWARE_CLASSES = [u'django.middleware.http.ConditionalGetMiddleware', u'django.contrib.sessions.middleware.SessionMiddleware', u'django.contrib.auth.middleware.AuthenticationMiddleware', u'django.contrib.messages.middleware.MessageMiddleware', u'django.middleware.csrf.CsrfViewMiddleware', u'django.middleware.locale.LocaleMiddleware', u'django.middleware.common.CommonMiddleware', u'cms.middleware.language.LanguageCookieMiddleware', u'cms.middleware.user.CurrentUserMiddleware', u'cms.middleware.page.CurrentPageMiddleware', u'cms.middleware.toolbar.ToolbarMiddleware', u'cms.middleware.utils.ApphookReloadMiddleware']
USE_I18N = True
THOUSAND_SEPARATOR = ','
CMS_LANGUAGES = {'default': {u'public': True, u'fallbacks': [u'en', u'de'], u'hide_untranslated': False, u'redirect_on_fallback': False}, 1: [{u'code': u'en', 'redirect_on_fallback': False, u'fallbacks': [u'de'], 'hide_untranslated': False, 'public': True, u'name': u'English'}, {u'code': u'de', 'redirect_on_fallback': False, u'fallbacks': [u'en'], 'hide_untranslated': False, 'public': True, u'name': u'German'}, {u'code': u'it', 'redirect_on_fallback': False, u'fallbacks': [u'en'], 'hide_untranslated': False, 'public': True, u'name': u'Italian'}, {u'code': u'zh-cn', 'redirect_on_fallback': False, u'fallbacks': [u'en'], 'hide_untranslated': False, 'public': True, u'name': u'Chinese Simplified'}], <class cms.utils.conf.VERIFIED at 0x10266c050>: True}
ALDRYN_BOOTSTRAP3_ICONSETS = 
LANGUAGE_COOKIE_NAME = 'django_language'
DEFAULT_INDEX_TABLESPACE = ''
LOGGING_CONFIG = 'logging.config.dictConfig'
SIGNING_BACKEND = 'django.core.signing.TimestampSigner'
TEMPLATE_LOADERS = 
FIRST_DAY_OF_WEEK = 0
WSGI_APPLICATION = None
TEMPLATE_DEBUG = False
X_FRAME_OPTIONS = 'SAMEORIGIN'
CSRF_COOKIE_NAME = 'csrftoken'
FORCE_SCRIPT_NAME = None
USE_X_FORWARDED_HOST = False
EMAIL_TIMEOUT = None
SECURE_SSL_HOST = None
DEFAULT_CHARSET = 'utf-8'
SESSION_COOKIE_SECURE = False
CSRF_COOKIE_DOMAIN = None
FILE_CHARSET = 'utf-8'
DEBUG = True
LANGUAGE_COOKIE_DOMAIN = None
DEFAULT_FILE_STORAGE = 'django.core.files.storage.FileSystemStorage'
INSTALLED_APPS = [u'django.contrib.contenttypes', u'django.contrib.auth', u'django.contrib.sessions', u'django.contrib.sites', u'django.contrib.staticfiles', u'djangocms_admin_style', u'django.contrib.admin', u'djangocms_helper.test_data', u'django.contrib.messages', u'cms', u'menus', u'sekizai', u'reversion', u'djangocms_text_ckeditor', u'djangocms_grid', u'filer', u'aldryn_bootstrap3', u'cms.test_utils.project.placeholderapp', u'treebeard', u'mptt']
LANGUAGES = 
USE_L10N = False
SECURE_HSTS_INCLUDE_SUBDOMAINS = False
STATICFILES_DIRS = 
PARLER_ENABLE_CACHING = False
PREPEND_WWW = False
SECURE_PROXY_SSL_HEADER = None
LANGUAGE_COOKIE_AGE = None
SESSION_COOKIE_HTTPONLY = True
PARLER_LANGUAGES = {u'default': {u'fallback': u'en', u'hide_untranslated': False}, 1: ({u'code': u'en', u'fallbacks': [u'de']}, {u'code': u'de', u'fallbacks': [u'en']}, {u'code': u'it', u'fallbacks': [u'en']}, {u'code': u'zh-cn', u'fallbacks': [u'en']})}
DEBUG_PROPAGATE_EXCEPTIONS = False
CSRF_COOKIE_AGE = 31449600
MONTH_DAY_FORMAT = 'F j'
LOGIN_URL = u'/admin/login/?user-login=test'
SESSION_EXPIRE_AT_BROWSER_CLOSE = False
SOUTH_TESTS_MIGRATE = True
USE_CMS = True
TIME_FORMAT = 'P'
AUTH_USER_MODEL = 'auth.User'
DATE_INPUT_FORMATS = 
AUTHENTICATION_BACKENDS = 'django.contrib.auth.backends.ModelBackend'
EMAIL_HOST_PASSWORD = u'********************'
PASSWORD_RESET_TIMEOUT_DAYS = u'********************'
SESSION_FILE_PATH = None
CACHE_MIDDLEWARE_ALIAS = 'default'
SESSION_SAVE_EVERY_REQUEST = False
NUMBER_GROUPING = 0
SESSION_ENGINE = 'django.contrib.sessions.backends.db'
CSRF_FAILURE_VIEW = 'django.views.csrf.csrf_failure'
CSRF_COOKIE_PATH = '/'
LOGIN_REDIRECT_URL = '/accounts/profile/'
CMS_CACHE_DURATIONS = {u'content': 0, u'menus': 0, u'permissions': 0}
DECIMAL_SEPARATOR = '.'
IGNORABLE_404_URLS = 
MIGRATION_MODULES = {}
TEMPLATE_STRING_IF_INVALID = ''
DISALLOWED_USER_AGENTS = 
LOGOUT_URL = '/accounts/logout/'
EMAIL_USE_TLS = False
FIXTURE_DIRS = 
EMAIL_HOST = 'localhost'
DATE_FORMAT = 'N j, Y'
MEDIA_ROOT = u'/var/folders/qf/m9l_j5_j3gz9vzfc3vkdj5300000gn/T/tmpKJkJHymedia'
DEFAULT_EXCEPTION_REPORTER_FILTER = 'django.views.debug.SafeExceptionReporterFilter'
ADMINS = 
FORMAT_MODULE_PATH = None
DEFAULT_FROM_EMAIL = 'webmaster@localhost'
MEDIA_URL = u'/media/'
DATETIME_FORMAT = 'N j, Y, P'
TEMPLATE_DIRS = 
SITE_ID = 1
CMS_PERMISSION = True
ALLOWED_INCLUDE_ROOTS = 
ALDRYN_BOOTSTRAP3_GRID_SIZE = 24
LOGGING = {}
SHORT_DATE_FORMAT = 'm/d/Y'
SECRET_KEY = u'********************'
TEMPLATES = [{u'OPTIONS': {u'context_processors': [u'django.contrib.auth.context_processors.auth', u'django.contrib.messages.context_processors.messages', u'django.template.context_processors.i18n', u'django.template.context_processors.csrf', u'django.template.context_processors.debug', u'django.template.context_processors.tz', u'django.template.context_processors.request', u'django.template.context_processors.media', u'django.template.context_processors.static', u'cms.context_processors.cms_settings', u'sekizai.context_processors.sekizai'], u'loaders': [u'django.template.loaders.filesystem.Loader', u'django.template.loaders.app_directories.Loader']}, u'DIRS': [u'/Users/divio/work/django-cms/cms/test_utils/project/templates/integration'], u'NAME': u'django', u'BACKEND': u'django.template.backends.django.DjangoTemplates'}]
TEST_RUNNER = 'django.test.runner.DiscoverRunner'
CACHE_MIDDLEWARE_KEY_PREFIX = u'********************'
SECURE_SSL_REDIRECT = False
TIME_ZONE = 'America/Chicago'
FILE_UPLOAD_MAX_MEMORY_SIZE = 2621440
EMAIL_BACKEND = u'django.core.mail.backends.locmem.EmailBackend'
BASE_APPLICATION = u'cms'
EMAIL_USE_SSL = False
TEMPLATE_CONTEXT_PROCESSORS = 
SESSION_COOKIE_AGE = 1209600
USE_ETAGS = False
LANGUAGES_BIDI = 
FILE_UPLOAD_TEMP_DIR = None
INTERNAL_IPS = 
STATIC_URL = u'/static/'
EMAIL_PORT = 25
USE_TZ = True
SHORT_DATETIME_FORMAT = 'm/d/Y P'
TEST_NON_SERIALIZED_APPS = []
PASSWORD_HASHERS = u'********************'
ABSOLUTE_URL_OVERRIDES = {}
LANGUAGE_COOKIE_PATH = '/'
CACHE_MIDDLEWARE_SECONDS = 600
EMAIL_SSL_CERTFILE = None
CSRF_COOKIE_HTTPONLY = False
DATETIME_INPUT_FORMATS = 
EMAIL_HOST_USER = ''

You're seeing this error because you have DEBUG = True in your
Django settings file. Change that to False, and Django will
display a standard page generated by the handler for this status code.
```

**Comment by czpython** [comment]
@vxsx 
So now the ``plugin`` parameter has been removed.
The url looks like ``/en/admin/cms/page/plugin/text_plugin/425/delete-on-cancel/``, as you can see this url includes the ID of the  text plugin in question.

When deleting the text plugin itself on cancel, just send a post request to that endpoint with the token.
When deleting child plugins, send all child plugins under the name ``child_plugins`` and the token.

Example (deleting text plugin)

```
curl 'http://127.0.0.1:8000/en/admin/cms/page/plugin/text_plugin/424/delete-on-cancel/' -d token=900ryuu37ieIvHQsaalUcwmXyA4
```

Example (deleting child plugins)
```
curl 'http://127.0.0.1:8000/en/admin/cms/page/plugin/text_plugin/424/delete-on-cancel/' -d token=900ryuu37ieIvHQsaalUcwmXyA4 -d child_plugins=425 -d child_plugins=426```

**Comment by vxsx** [comment]
seems to work @czpython

File: djangocms_text_ckeditor/static/djangocms_text_ckeditor/js/cms.ckeditor.js [hunk 1]
```diff
@@ -84,7 +84,11 @@ $(document).ready(function () {
                 that.editor.resize('100%', win.CMS.$('.cms-modal-frame').height() - 70);
                 this.editor.execCommand('maximize');
                 win.CMS.API.Helpers.addEventListener('modal-maximized modal-restored', function (e, payload) {
-                    that.editor.resize('100%', win.CMS.$('.cms-modal-frame').height() - 70);
+                    try {
+                        that.editor.resize('100%', win.CMS.$('.cms-modal-frame').height() - 70);
+                    } catch (e) {
+                        // sometimes throws errors if modal with text plugin is closed too fast
```
Comment (FinalAngel) [inline djangocms_text_ckeditor/static/djangocms_text_ckeditor/js/cms.ckeditor.js]: we might want to trigger a warning? at least in the console?

Comment (vxsx) [inline djangocms_text_ckeditor/static/djangocms_text_ckeditor/js/cms.ckeditor.js]: not really. we just want it not to throw an error, this happens if editor doesn't exist anymore by the time event is fired
warning won't really be helpful

Comment (FinalAngel) [inline djangocms_text_ckeditor/static/djangocms_text_ckeditor/js/cms.ckeditor.js]: alright :+1:

**Comment by czpython** [comment]
Merged into develop branch. Leave this PR open for now.

**Pull request merged**